### PR TITLE
fix TreeBehavior not returning false on moveUp/moveDown accordingly

### LINKED ISSUE #16091: TreeBehavior::moveUp does not return false for the root element
### Description

Moving the root element up does not return false as promised by the [documentation](https://api.cakephp.org/3.10/class-Cake.ORM.Behavior.TreeBehavior.html#moveUp()).

```
        $entity = $this->Areas->get($id);
        $moved = $this->Areas->moveUp($entity, 1);
```
$moved is an entity object.

The entity fields are the same before and after the call to moveUp:
```
    [id] => 1
    [name] => ...name...
    [parent_id] => null
    [lft] => 1
    [rght] => 190
```

Logs:

```
2021-11-01 10:26:53 Debug: duration=0 rows=0 BEGIN
2021-11-01 10:26:53 Debug: duration=0 rows=0 SELECT `Area`.`lft` AS `Area__lft`, `Area`.`rght` AS `Area__rght` FROM `areas` `Area` WHERE ((`parent_id`) IS NULL AND (`rght`) < 1) ORDER BY `lft` DESC LIMIT 1 OFFSET 0
2021-11-01 10:26:53 Debug: duration=0 rows=0 SELECT `Area`.`lft` AS `Area__lft`, `Area`.`rght` AS `Area__rght` FROM `areas` `Area` WHERE ((`parent_id`) IS NULL AND (`rght`) < '1') ORDER BY `lft` ASC LIMIT 1
2021-11-01 10:26:53 Debug: duration=0 rows=0 COMMIT
```

### CakePHP Version

3.10.0

### PHP Version

7.4

### PR DESCRIPTION
**Pull request by LordSimal**
fixes #16091

### CODE REVIEW

**Comment by LordSimal** [comment]
It seemed like the TreeBehavior returned the given entity if no left/right node was found.
This does not correspond to the behaviour described in the PHPDoc/Cookbook.

Therefore I adjusted the return value in that case to false and fixed the tests.

File: tests/TestCase/ORM/Behavior/TreeBehaviorTest.php [hunk 1]
```diff
@@ -427,7 +427,7 @@ public function testMoveLeaf(): void
         $table = $this->getTableLocator()->get('MenuLinkTrees');
         $table->addBehavior('Tree', ['scope' => ['menu' => 'main-menu']]);
         $node = $table->moveUp($table->get(5), 1);
-        $this->assertEquals(['lft' => 6, 'rght' => 7], $node->extract(['lft', 'rght']));
+        $this->assertEquals(false, $node);
```
Comment (markstory) [inline tests/TestCase/ORM/Behavior/TreeBehaviorTest.php]: Should these be updated to `assertSame()` or `asertFalse()` so we don't have loose comparisons to null passing?

**Comment by saeideng** [comment]
this is behavior change and will  breaks current programs
this behavior was from `3.0.0`
even if this was acceptable, this should target on 4.4 IMO
I'm not fine with surprising change in bugfix

or other suggestion : 
having new methods , maybe : `moveUpOrFail()` / `moveUpOrFalse()` ,

File: tests/TestCase/ORM/Behavior/TreeBehaviorTest.php (hunk 1 above)
Comment (LordSimal) [reply tests/TestCase/ORM/Behavior/TreeBehaviorTest.php]: I will do that now

**Comment by markstory** [comment]
> this is behavior change and will breaks current programs
> this behavior was from 3.0.0
> even if this was acceptable, this should target on 4.4 IMO
> I'm not fine with surprising change in bugfix

These are good points. The current behavior of `moveUp` has been in place since 2015 (c3b2f8bc1de). The lower risk change would be to update the documentation to match the long standing behavior, or change the failures to be exceptions. Introducing exceptions is within our backwards compatibility guidelines but doesn't feel great as it will still break user land code.

> or other suggestion :
> having new methods , maybe : moveUpOrFail() / moveUpOrFalse()

If we introduce new methods I would prefer exceptions over mixed return types. That matches with the rest of the ORM method signatures more.

**Comment by ADmad** [comment]
I agree we should just update the docs. I wouldn't want an exception either. Trying to move up a record which is already at the top is essentially a no-op and doesn't cause any actual problem which would warrant a *hard* error.

**Comment by saeideng** [comment]
> I agree we should just update the docs

me too

and maybe  instead of returning `false` we can throw an exception in `5.x`  (when `$number < 1` )

then we can add return type to these methods in `5.x`
it means removing `false` from return type
I'm not sure about throwing an exception yet
```php
    public function moveUp(EntityInterface $node, $number = 1)
    {
        if ($number < 1) {
            return false;
        }
```

**Comment by saeideng** [comment]
right now we have an exception when record not found

https://github.com/cakephp/cakephp/blob/ee64decafc78010747a251a36767af5a4ba5dc2e/src/ORM/Behavior/TreeBehavior.php#L608-L611

**Comment by LordSimal** [comment]
So the current plan is to

- keep the logic as it is (not returning false and therefore not breaking already present logic based on it)
- adjust the PHPDoc to represent that (basically remove the false return option)

I re-checked the Cookbook and couldn't find any mention of that.

Only the API doc https://api.cakephp.org/4.3/class-Cake.ORM.Behavior.TreeBehavior.html#moveDown() mentions the false return type but I guess this gets generated from the source, right?

The actual change in behaviour will happen in `5.x`.

**Comment by markstory** [comment]
> Only the API doc ... mentions the false return type but I guess this gets generated from the source, right?

Yes the API docs are generated from the source.

> So the current plan is to ...

Yes.

**Comment by LordSimal** [comment]
I made a separate PR for the doc change because this change will still be necessary when `5.x` is going to come.

Don't know how you would like to handle this PR.

**Comment by othercorey** [comment]
Closing in favor of the updating the docblocks.

**Pull request closed without merging**